test/extensions/filters/http/ext_proc/streaming_integration_test.cc

```diff
@@ -235,30 +386,248 @@ TEST_P(StreamingIntegrationTest, GetAndProcessBufferedResponseBody) {
+          // Save all the chunks in a buffer so that we can calculate a hash.
+          allData.add(body_req.response_body().body());
+          ProcessingResponse body_resp;
+          body_resp.mutable_response_body();
+          stream->Write(body_resp);
+        } while (!body_req.response_body().end_of_stream());
+
+        processor_response_hash_ = HashUtil::xxHash64(allData.toString());
+        ASSERT_EQ(total_response_size, response_size);
       });
 
   initializeConfig();
   HttpIntegrationTest::initialize();
   Http::TestRequestHeaderMapImpl headers;
   HttpTestUtility::addDefaultHeaders(headers);
+  // This magic header tells the "autonomous upstream" to send back a response
+  // of the specified size.
   headers.addCopy(LowerCaseString("response_size_bytes"), response_size);
   sendGetRequest(headers);
 
   ASSERT_TRUE(client_response_->waitForEndStream());
   EXPECT_TRUE(client_response_->complete());
   EXPECT_THAT(client_response_->headers(), Http::HttpStatusIs("200"));
   EXPECT_EQ(client_response_->body().size(), response_size);
+  EXPECT_EQ(processor_response_hash_, HashUtil::xxHash64(client_response_->body()));
+}
+
+// Send a large HTTP POST, and expect back an equally large reply. Stream both and verify
+// that we got back what we expected. The processor itself must be written carefully
+// because once the request headers are delivered, the request and response body
+// chunks and the response headers can come in any order.
+TEST_P(StreamingIntegrationTest, PostAndProcessStreamBothBodies) {
+  const uint32_t send_chunks = 10;
+  const uint32_t chunk_size = 11000;
+  uint32_t request_size = send_chunks * chunk_size;
+  uint32_t response_size = 1700000;
+
+  test_processor_.start(
+      [this, request_size,
+       response_size](grpc::ServerReaderWriter<ProcessingResponse, ProcessingRequest>* stream) {
+        ProcessingRequest header_req;
+        ASSERT_TRUE(stream->Read(&header_req));
+        ASSERT_TRUE(header_req.has_request_headers());
+        ProcessingResponse header_resp;
+        header_resp.mutable_request_headers();
+        stream->Write(header_resp);
+
+        bool saw_response_headers = false;
+        bool saw_request_eof = false;
+        bool saw_response_eof = false;
+        ProcessingRequest message;
+        uint32_t total_request_size = 0;
+        uint32_t total_response_size = 0;
+        Buffer::OwnedImpl allResponseData;
+        Buffer::OwnedImpl allRequestData;
+
+        do {
+          ProcessingResponse response;
+          ASSERT_TRUE(stream->Read(&message));
+          if (message.has_response_headers()) {
+            // Expect only one response headers message, with a good status.
+            EXPECT_FALSE(saw_response_headers);
+            EXPECT_THAT(message.response_headers().headers(),
+                        SingleProtoHeaderValueIs(":status", "200"));
+            saw_response_headers = true;
+            response.mutable_response_headers();
+
```

Comment:
Nit: quite a few places in this PR have this gratuitous blank line at the end of an `if` block. Prefer compact style as per rest of code base.

Reply:
I removed many of these. I still have blank lines in a few places where I think it clarifies things for me, but not immediately around conditions, and I'll try to restrain my natural tendency for whitespace in the future.